Add support for Cursor type override for federation compatibility (#1231)
block/elasticgraph: config/site/src/guides/customizing-the-graphql-schema.md

The standard GraphQL scalars (`Boolean`, `Float`, `ID`, `Int`, `String`) and the root `Query` type cannot be renamed
this way.

### Federation Compatibility: Overriding `Cursor` to `String`

When composing an ElasticGraph subgraph into a federated supergraph alongside other subgraphs that use `String` for
cursor fields (as the [Relay GraphQL Cursor Connections Specification](https://relay.dev/graphql/connections.htm) permits),
federation composition may fail with a type incompatibility error. ElasticGraph uses a dedicated `Cursor` scalar type
for cursor fields by default, which provides better type safety and documentation but can cause conflicts.

To resolve this, override the `Cursor` type to `String`:

```ruby
# Within `ElasticGraph::Local::RakeTasks.new { ... }` in your `Rakefile`:
tasks.type_name_overrides = {Cursor: "String"}

> ```

Reviewer: ```suggestion
{% include copyable_code_snippet.html language="ruby" data="schema_customization_rake_tasks.snippets.Rakefile.cursor_type_override" %}
```

As per our [guiding principles](https://block.github.io/elasticgraph/guides/guiding-principles/#codebase-principles), "We aim to validate all snippets and code examples at this website", and the way we do that is to pull in snippets that are validated on CI instead of having inline snippets which could silently break as EG's APIs evolve.

The standard GraphQL scalars (`Boolean`, `Float`, `ID`, `Int`, `String`) and the root `Query` type cannot be renamed
this way.

### Federation Compatibility: Overriding `Cursor` to `String`

When composing an ElasticGraph subgraph into a federated supergraph alongside other subgraphs that use `String` for

> cursor fields (following the [Relay GraphQL Cursor Connections Specification](https://relay.dev/graphql/connections.htm)),

Reviewer: ```suggestion
cursor fields (as the [Relay GraphQL Cursor Connections Specification](https://relay.dev/graphql/connections.htm) permits),
```

ElasticGraph's `Cursor` type already follows the Relay spec. The [Relay spec](https://relay.dev/graphql/connections.htm#sel-EAHFHDDAADFCa4iH) specifically allows custom scalars to be used for cursors instead of `String`:

> An “Edge Type” must contain a field called cursor. This field must return a type that serializes as a String; this may be a String, a Non-Null wrapper around a String, a custom scalar that serializes as a String, or a Non-Null wrapper around a custom scalar that serializes as a String.
>
> Whatever type this field returns will be referred to as the cursor type in the rest of this spec.

I read the Relay spec closely when implementing pagination in ElasticGraph and chose to use a custom scalar type based on that.

federation composition may fail with a type incompatibility error. ElasticGraph uses a dedicated `Cursor` scalar type
for cursor fields by default, which provides better type safety and documentation but can cause conflicts.

To resolve this, override the `Cursor` type to `String`:

```ruby
# Within `ElasticGraph::Local::RakeTasks.new { ... }` in your `Rakefile`:
tasks.type_name_overrides = {Cursor: "String"}
```

This configuration causes ElasticGraph to:
- Skip registration of the `Cursor` scalar (avoiding duplicate type definitions)
- Use `String` for all cursor-related fields (`PageInfo.startCursor`, `PageInfo.endCursor`, `Edge.cursor`)
- Use `String` for pagination arguments (`before`, `after`)

Internally, ElasticGraph will lazily decode cursor strings when pagination logic requires access to the decoded cursor
structure. This decoding is transparent—pagination continues to work exactly as before, including forward and backward
pagination, cursor validation, and aggregation pagination.

> 

Reviewer: ```suggestion
```

{: .alert-note}
**Note**{: .alert-title}
The `Cursor` scalar and `String` are semantically identical on the wire—both are opaque base64-encoded strings. The
only difference is that `Cursor` provides more expressive type information in the GraphQL schema. Using `String` for

> cursor fields is fully compatible with the Relay specification and is the common convention in most GraphQL implementations.

Reviewer: > is the common convention in most GraphQL implementations.

Funny, until today I thought the common convention was to use a `Cursor` type like ElasticGraph uses. I checked the schemas of a few popular GraphQL APIs and it seems like you're right that `String` is the common convention. I thought I was following the standard convention when I chose to define a `Cursor` scalar. TIL!
